subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+		data_fifo_empty(handle->thread.msg_rx);
+	}
+
+	if (handle->thread.msg_tx != NULL) {
+		data_fifo_empty(handle->thread.msg_tx);
+	}
+
+	/*
+	 * How to return all the data to the slab items?
+	 * Test the semaphore and wait for it to be zero.
+	 */
+
+	k_thread_abort(handle->thread_id);
+
+	LOG_DBG("Closed module %s", handle->name);
+
+	/* Clean up the handle */
+	memset(handle, 0, sizeof(struct audio_module_handle));
+
+	return 0;
+};
+
+/**
+ * @brief  Function to reconfigure a module.
+ */
+int audio_module_reconfigure(struct audio_module_handle *handle,
+			     struct audio_module_configuration *configuration)
+{
+	int ret;
+
+	if (handle == NULL || configuration == NULL) {
+		LOG_DBG("Module input parameter error");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED || handle->state == AMOD_STATE_RUNNING) {
+		LOG_DBG("Module %s in an invalid state, %d, for setting the configuration",
+			handle->name, handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->description->functions->configuration_set != NULL) {
+		ret = handle->description->functions->configuration_set(
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_DBG("Set configuration for module %s send failed, ret %d", handle->name,
+				ret);
+			return ret;
+		}
+	} else {
+		LOG_DBG("Reconfiguration for module %s has no set configuration function",
+			handle->name);
+	}
+
+	if (handle->state != AMOD_STATE_CONFIGURED) {
+		handle->previous_state = handle->state;
+	}
+
+	handle->state = AMOD_STATE_CONFIGURED;
+
+	return 0;
+};
+
+/**
+ * @brief  Function to get the configuration of a module.
+ */
+int audio_module_configuration_get(struct audio_module_handle *handle,
+				   struct audio_module_configuration *configuration)
+{
+	int ret;
+
+	if (handle == NULL || configuration == NULL) {
+		LOG_DBG("Module input parameter error");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED) {
+		LOG_DBG("Module %s in an invalid state, %d, for getting the configuration",
+			handle->name, handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->description->functions->configuration_get != NULL) {
+		ret = handle->description->functions->configuration_get(
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_DBG("Get configuration for module %s failed, ret %d", handle->name,
+				ret);
+			return ret;
+		}
+	} else {
+		LOG_DBG("Get configuration for module %s has no get configuration function",
+			handle->name);
+	}
+
+	return 0;
+};
+
+/**
+ * @brief Function to connect two modules together.
+ *
+ */
+int audio_module_connect(struct audio_module_handle *handle_from,
+			 struct audio_module_handle *handle_to)
+{
+	struct audio_module_handle *handle;
+
+	if (handle_from == NULL || handle_to == NULL) {
+		LOG_DBG("Invalid parameter for the connection function");
+		return -EINVAL;
+	}
+
+	if (handle_from->description->type == AMOD_TYPE_UNDEFINED ||
```

Comment:
This works, but I suggest testing the combinations which we know works; incase the number of types are expanded later on.

Reply:
Can we discuss this?

Reply:
After discussion: Will change to check for valid states.